source/ConditionSet.cpp

```diff
@@ -233,78 +133,215 @@ ConditionSet::ConditionSet(const DataNode &node)
 void ConditionSet::Save(DataWriter &out) const
 {
-	for(const Expression &expression : expressions)
-		expression.Save(out);
+	// Default should be AND, so if it is, then just write the subsets.
+	// If this condition got optimized beyond OP_AND, then re-add the OP_AND by writing the current condition in full.
+	if(expressionOperator == OP_AND)
+		for(const auto &child : children)
+		{
+			child.SaveSubset(out);
+			out.Write();
+		}
+	else
+		SaveSubset(out);
+}
+
+
+
+void ConditionSet::SaveChild(int childNr, DataWriter &out) const
+{
+	const ConditionSet &child = children[childNr];
+	bool needBrackets = child.children.size() > 0;
+
+	if(needBrackets)
+		out.WriteToken("(");
+	children[childNr].SaveSubset(out);
+	if(needBrackets)
+		out.WriteToken(")");
+}
+
+
 
-	for(const ConditionSet &child : children)
+// Save a subset of conditions, by writing out tokens (without a newline).
+void ConditionSet::SaveSubset(DataWriter &out) const
+{
+	string opTxt = "";
+	auto it = find_if(CS_TOKEN_CONVERSION.begin(), CS_TOKEN_CONVERSION.end(),
+		[this](const std::pair<const string, ConditionSet::ExpressionOp> &e) {
+			return e.second == expressionOperator;
+		});
+	if(it != CS_TOKEN_CONVERSION.end())
+		opTxt = it->first;
+
+	switch(expressionOperator)
 	{
-		out.Write(child.isOr ? "or" : "and");
+	case OP_INVALID:
+		out.WriteToken("never");
+		break;
+	case OP_VAR:
+		out.WriteToken(conditionName);
+		break;
+	case OP_LIT:
+		out.WriteToken(literal);
+		break;
+	case OP_ADD:
+	case OP_SUB:
+	case OP_MUL:
+	case OP_DIV:
+	case OP_MOD:
+	case OP_EQ:
+	case OP_NE:
+	case OP_LE:
+	case OP_GE:
+	case OP_LT:
+	case OP_GT:
+		if(children.empty())
+		{
+			out.WriteToken("never");
+			break;
+		}
+		SaveChild(0, out);
+		for(unsigned int i = 1; i < children.size(); ++i)
+		{
+			out.WriteToken(opTxt);
+			SaveChild(i, out);
+		}
+		break;
+	case OP_AND:
+	case OP_OR:
+		out.Write(opTxt);
 		out.BeginChild();
+		for(const auto &child : children)
 		{
-			child.Save(out);
+			child.SaveSubset(out);
+			out.Write();
 		}
 		out.EndChild();
-	}
+		break;
+	case OP_NOT:
+	case OP_HAS:
+		if(children.empty())
+		{
+			out.WriteToken("never");
+			break;
+		}
+		out.WriteToken(opTxt);
+		SaveChild(0, out);
+		break;
+	default:
+		out.WriteToken("never");
+		break;
+	};
 }
 
 
 
 void ConditionSet::MakeNever()
 {
-	// Add the equivalent "never" condition, `"'" != 0`.
-	// TODO: change the ConditionSet to contain "false" literal, instead of a comparison that should return "false".
-	// TODO: Validate if condition-names are valid. (The "'" character by itself should not pass as a valid condition.)
-	Add("has", "'");
+	children.clear();
+	expressionOperator = OP_LIT;
+	literal = 0;
 }
 
 
 
 // Check if there are any entries in this set.
+// Invalid ConditionSets are also considered empty.
 bool ConditionSet::IsEmpty() const
 {
-	return expressions.empty() && children.empty();
+	return
+		(expressionOperator == OP_AND && children.size() == 0) ||
```

Comment:
AND and OR had the same handling when being saved, but not here.

Reply:
The default top-level operator is OP_AND. So when absolutely nothing got parsed (because the complete expression was empty), then we will have an OP_AND operand and no children under it.

I could add a default OP_EMPTY operator for clarity, but maybe it is sufficiently clear if I just add a comment here?